ansible_base/rbac/permission_registry.py

```diff
@@ -22,43 +22,59 @@
+        self.model_name = model._meta.model_name
+        self.app_label = model._meta.app_label
+        if parent_field_name == self.model_name:
+            # model can not be its own parent
+            self.parent_field_name = None
+        else:
+            self.parent_field_name = parent_field_name
+        self.allow_object_roles = allow_object_roles
+        self.model = model
+
+
 class PermissionRegistry:
     def __init__(self):
-        self._registry = set()  # model registry
-        self._name_to_model = dict()
-        self._parent_fields = dict()
+        self._registry = dict()  # model registry
         self._managed_roles = dict()  # code-defined role definitions, managed=True
         self.apps_ready = False
         self._tracked_relationships = set()
         self._trackers = dict()
 
-    def register(self, *args, parent_field_name='organization'):
+    def register(self, *args, **kwargs):
         if self.apps_ready:
             raise RuntimeError('Cannot register model to permission_registry after apps are ready')
-        for cls in args:
-            if cls not in self._registry:
-                self._registry.add(cls)
-                model_name = cls._meta.model_name
-                if model_name in self._name_to_model:
-                    raise RuntimeError(f'Two models registered with same name {model_name}')
-                self._name_to_model[model_name] = cls
-                if model_name != 'organization':
-                    self._parent_fields[model_name] = parent_field_name
+        for model in args:
+            if model._meta.model_name not in self._registry:
+                info = ModelPermissionInfo(model, **kwargs)
+                self._registry[info.model_name] = info
+            elif self._registry[model._meta.model_name] is model:
+                logger.debug(f'Model {model._meta.model_name} registered to permission registry more than once')
             else:
-                logger.debug(f'Model {cls._meta.model_name} registered to permission registry more than once')
+                raise RuntimeError(f'Two models registered with same name {model._meta.model_name}')
+
+    def get_info(self, obj: Union[ModelBase, Model]) -> ModelPermissionInfo:
+        return self._registry[obj._meta.model_name]
 
     def track_relationship(self, cls, relationship, role_name):
         self._tracked_relationships.add((cls, relationship, role_name))
 
     def get_parent_model(self, model) -> Optional[type]:
-        model = self._name_to_model[model._meta.model_name]
-        parent_field_name = self.get_parent_fd_name(model)
-        if parent_field_name is None:
+        info = self._registry[model._meta.model_name]
+        if info.parent_field_name is None:
             return None
-        return model._meta.get_field(parent_field_name).related_model
+        return model._meta.get_field(info.parent_field_name).related_model
```

Comment:
Is it possible to register a model with a parent but not the parent?

Reply:
Yes, very much so, `InstanceGroup` does not have a related organization. So you can have an object-level role, but no organization association. This call was made in fairly recent development, but before any work on the new RBAC system had started.

Even more interestingly, the `Instance` model is the one case where not being organization-scoped is truly defensible (an instance is a property of the cluster...), but object-level roles might also not make sense. It's more likely that you would _only_ want to delegate permissions with system-level roles. Because you're kind of either managing an AWX cluster or you're not... it's 1 thing, and sub-dividing it doesn't make a ton of sense.